### DISCUSSION TITLE
Tasmota Script and ufs - write Array to open File

### DISCUSSION THREAD

**Discussion by thc08** [arendst/Tasmota#20878] [Q&A] [answered]
Hello.
Can someone explain to me (preferably with a working one-liner) how to write an array to an open file in ufs? Normal script, no Berry.
"fwa(array fr (a)) writes array to open file with fr ....."

Wont work for me ,  -> var not found: fwa(mcon fr)

Thanks

**Comment by HofeBY** [comment]
In Script without Berry using a ESP32 with internal Flash as Filesystem:

`>D`
rssi=0 ; WiFi strength
AOAZ=1234.3333 ;Variable change in Section >T
ATAZ=5678.1234 ;Variable change in Section >T

`>B`

`>T`
rssi=WiFi#rssi
AOAZ=AMT681#A1_8_0 ; consumption Electricity Meter
ATAZ=AMT681#A2_8_0 ; excess Electricity Meter

`>S`
; Write messurements every 60sec into Logfile if activated
;if upsecs%60==0
;then ; maybe replace by {
fr=fo("/FName-"+s(0year)+"B.csv" 2)
res=fw(s(0year)+s(0month)+s(0day)+";"+s(4AOAZ)+";"+s(4ATAZ) fr)
res=fw(";"+s(0rssi)+"\r\n" fr)
fc(fr)
;endif ; maybe replace by }

`>W`
;These lines are displayed on http site.
Timestamp: %tstamp%h {m}
%dvnm% WLAN IP: %lip% {m}%rssi% %%
AOAZ = %4AOAZ% Bezug A-Zähler
ATAZ = %4ATAZ% Ueberschuss A-Zähler
ATAZ = %ATAZ% ; without number of digits ->2

After starting look into File space of Tasmota.

**Reply by sfromis** [reply]
To copy/paste such code, I can recommend having lines with triple backtick before/after. Becomes more readable, and avoid "tricks" like "replace with ;" to defeat other automatic formatting.

**Reply by sfromis** [reply]
Having those triple-backtick lines is the key to make the formatting work out nicely.
https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#fenced-code-blocks

**Comment by thc08** [comment]
@HofeBY

Hello, thank you first of all. Unfortunately does not work. No file is created and data is logically not written either.
No error message with Weblog 4. Surprisingly, the example from the docs (sensor logging) works. 
If I use "fwa", the same thing happens as with your script, no file, no error message.
It is an ESP32 S2 with Tasmota 13.3.0, built with the Tasmocompiler for S2. 

Additionally the following define:

#ifndef LARGE_ARRAYS 
#define LARGE_ARRAYS 
#endif 
#ifndef 
#USE_SCRIPT_WEB_DISPLAY 
#define 
#USE_SCRIPT_WEB_DISPLAY 
#endif 
#ifndef 
#USE_SCRIPT_GLOBVARS 
#define 
#USE_SCRIPT_GLOBVARS 
#endif 
#ifndef 
#USE_UFILESYS 
#define USE_UFILESYS 
#endif 
#define UFSYS_SIZE 16384 
#define USE_FEXTRACT 
#define USE_SCRIPT_FATFS_EXT ; Test if it is missing 
#define SCRIPT_STRIP_COMMENTS

Could it be that I'm missing something?

Thanks and Greetings

**Reply by sfromis** [reply]
That `#define` format looks rather bad. `#define/#ifndef` and the name has to be on the same line, the name after `#define/#ifndef` should not start with `#`, and the comment delimiter is `//` instead of `;`. The syntax to use there is very different from the scripter.

**Reply by thc08** [reply]
@sfromis 
Hello, yes i now. 
Was a problem with Editor, Handy and me :-)
The comment is only for here.
Thanks

**Comment by HofeBY** [comment]
Maybe you want to use my user_config_override.h as an example for custom compiled version to your ESP32 S2.

[[github.com/HofeBY/Tasmota-Firmware](https://github.com/HofeBY/Tasmota-Firmware)](url)

sorry for german example, but you can adopt it to your local needs.

**Reply by thc08** [reply]
@HofeBY 

Thanks
German is fine :-)

Wont work, I'll get another Esp32. And try Image from you. Unbelievable, I don't understand it.
Especially since the example from the docs works.
I'm running out of time right now, but I'll report back when I've solved the problem. 

Gruss!

**Reply by sfromis** [reply]
If you're using ESP32, the Berry language is also a pretty good way of handling data, and "less arcane" in the syntax for doing things.

**Reply by HofeBY** [reply]
@thc08
update to TM13.4.0.2 with several more ESP types.

**Reply by thc08** [reply]
@sfromis 

Hello,

Yes i now. But Berry is to much/hard for me.
I have around 30 Devices, mixed Esp/Esp32, so normal Script worked on all.

Grertings ..

**Reply by sfromis** [reply]
I'd say that the many intricacies of using the scripter is quite a bit harder to get used to than Berry. But of course, if you have already climbed the learning curve, it can make sense to stay with that, even when still having to ask questions about relatively basic things.

**Comment by HofeBY2** [comment]
As I know, Berry is not possible under ESP8285 and ESP8266. So I decided primary to share Script for all ESP Tasmota's.

**Reply by sfromis** [reply]
My suggestion was not directed at you, but at thc08 talking about using ESP32.

**Comment by pkkrusty** [comment]
Your variable of A180 seems like it could be causing trouble. Maybe change to all text characters. 

Also array maybe isn't defined correctly. Needs to be m:con=0 19 (or however large the array needs to be)

**Reply by HofeBY** [reply]
Why A180 as a variable isn´t possible ?
Wouldn't be so good if you can't number variables by function. This is possible in almost all interpreter languages.

Earlier Tasmota versions correctly differentiate between the variables "A180" and "A280", newer Tasmota versions mix the content, which is why I have to rewrite my applications.
Can be found more or less well.

OK, renaming it to AOEZ ( A for Meter; O for One; E for Eight; Z for Zero)

did i use m:con as a variable ? ->NO
OK changing to ATEZ ( A for Meter; T for Two; E for Eight; Z for Zero)

OK for you ?

**Comment by pkkrusty** [comment]
You write: 
```
Wont work for me , -> var not found: fwa(mcon fr)
```
You haven't defined an array, so you get a variable not defined error. An array is defined like 

`m:con=0 19` where 19 is the length of the array. It looks like you are just wanting to write regular variables into your file.  In that case, you don't need fwa, you just use fw. Sensor example in docs uses fw, not fwa.

**Reply by pkkrusty** [reply]
Also, I think you can't create a string within the fw command.  You need to define the string before and then just write the string variable with the fw() command. So 
```
>D
rssi=0 ; WiFi strength
AOAZ=1234.3333 ;Variable change in Section >T
ATAZ=5678.1234 ;Variable change in Section >T
string1=""
file1=""
>B

>T
rssi=WiFi#rssi
AOAZ=AMT681#A1_8_0 ; consumption Electricity Meter
ATAZ=AMT681#A2_8_0 ; excess Electricity Meter

>S
; Write messurements every 60sec into Logfile if activated
;if upsecs%60==0
;then ; maybe replace by {
file1="/FName-"+s(0year)+"B.csv"
fr=fo(file1 2)
string1=s(0year)+s(0month)+s(0day)+";"+s(4AOAZ)+";"+s(4ATAZ)
res=fw(string1 fr)
string1=";"+s(0rssi)+"\r\n"
res=fw(string1 fr)
fc(fr)
;endif ; maybe replace by }

>W
;These lines are displayed on http site.
Timestamp: %tstamp%h {m}
%dvnm% WLAN IP: %lip% {m}%rssi% %%
AOAZ = %4AOAZ% Bezug A-Zähler
ATAZ = %4ATAZ% Ueberschuss A-Zähler
ATAZ = %ATAZ% ; without number of digits ->2
```

**Reply by pkkrusty** [reply]
With ESP32, you have tons of RAM, so you can make max string length longer and fit everything in one line instead of doing it in two writes.

**Reply by thc08** [reply]
@pkkrusty 

Hello.

Arrays are defined.

m:p:mcon:0 24

23:08:37.482 CMD: script?mcon
23:08:37.490 RSL: RESULT = {"script":{"mcon":[0,0,41.04,0,0,0,0,0,0,0,0,0,100,100,2.87,0,0,0,0,0,0,0,0,0]}}

I would like to write this to the log file, it should be an array. [x,x,.....]

I will test the script you posted. Unfortunately my time is extremely limited at the moment.

Greetings

**Reply by gemu2015** [reply]
array example.
fwa() stores numeric array as tab delimited strings, string decimals are set from dp() cmd
fra() restores them

```
>D
M:array=0 8
cnt=0
fres=0

>B
; set decimal precision (default = 2)
dp(0 5)
; call read array
=#readarray

; read array
#readarray
fres=fo("/array.txt" r)
if fres>=0 {
fra(array fres)
fc(fres)
print array loaded
}

; preset array
#preset
array={1 2 3 4 5 6 7 8}
; and write
fres=fo("/array.txt" w)
if fres>=0 {
fwa(array fres)
fc(fres)
print array saved
}

#show
if wm==0 {
for cnt 1 array[-1] 1
wcs array[%0cnt%] = %array[cnt]%
wcs <br>
next
}

>W
; show array in web ui
%=#show


```

**Reply by thc08** [reply]
@gemu2015

Hello
Thanks a lot for this example.
I think you have tested this example, so there must be an Error on OSI8 :-)
The W Sektion work, Show all.
Array.txt was written with 0 Bytes.
Tested with Tasmota 13.3 ,13.4 on esp32s2 esp32c3 wemos D1 Mini.

Thanks ...

**Reply by gemu2015** [reply]
when you call #preset array in console the array is written.
script>=#preset

**Reply by thc08** [reply]
@gemu2015 

Hello, yes array.txt is written with 0 bytes.
#readarry , wont read array.txt
No Error in the weblog (level 4)

Gruss

**Reply by gemu2015** [reply, marked as answer]
sorry, if forgot that you need extended file system support
#define USE_SCRIPT_FATFS_EXT

**Reply by thc08** [reply]
Hello Gemu, thank you very very much.

Of course it worked, unfortunately my test images for the Esp's were compiled without ' #define USE_SCRIPT_FATFS_EXT' :-( 
Then the Tasmota Docs are not correct at this point, no requirement for fatfs_ext, for this command (and probably for others).
I know myself That's not the case, should I report this somewhere so that a correction can be made? 

I still have a question regarding lfw("fname" payload limit) e.g. script>lfw("/test.log" cts(tstamp 1) 256) or script>lfw("/test.log" foobar 256) that works so far. 
Only when the limit is reached nothing is written and of course the first line is not deleted,
the esp becomes extremely slow and then restarts with an exception.

Greetings!

**Reply by gemu2015** [reply]
you are very welcome to improve the docs. everyone can do this, just use the pencil in the upper right.

lfw has a limitation which is also not in the docs. when the limit is reached the file needs to be copied and therefore the free file space must be at least double the file size, better much more.

**Reply by thc08** [reply]
Hello,
i will try my best with the documentation.
However, I am rather unsuitable. :-)

I'll test it out with lfw.

Thank you again.
